vortex-buffer/src/buffer.rs

```diff
@@ -0,0 +1,218 @@
+}
+
+impl<T: ByteSource + Sync + Send + 'static> ByteOwner for T {
+    #[inline(always)]
+    fn as_any(self: Arc<Self>) -> Arc<dyn Any + Sync + Send> {
+        self
+    }
+
+    #[inline(always)]
+    fn into_mut(self: Arc<Self>) -> Result<BufferMut, Arc<dyn ByteOwner>>
+    where
+        Self: Sized,
+    {
+        <T as ByteSource>::into_mut(self)
+    }
+}
+
+impl Buffer {
+    /// Creates a new empty `Buffer`.
+    pub fn empty() -> Self {
+        Self::from_owner(&[0u8; 0][..])
+    }
+
+    /// Creates `Buffer` from a [`ByteOwner`] + [`ByteSource`].
+    pub fn from_owner(owner: impl ByteSource + ByteOwner) -> Self {
+        let owner = Arc::new(owner);
+        let slice = owner.as_bytes();
+        let ptr = slice.as_ptr();
+        let length = slice.len();
+        Self { owner, ptr, length }
+    }
+
+    /// Creates `Buffer` from an `Arc<ByteSource + ByteOwner>`.
+    pub fn from_owner_arc(arc: Arc<impl ByteSource + ByteOwner>) -> Self {
+        let slice = arc.as_bytes();
+        let ptr = slice.as_ptr();
+        let length = slice.len();
+        Self {
+            owner: arc,
+            ptr,
+            length,
+        }
+    }
+
+    /// Extract and downcast the owner of the bytes.
+    /// TODO(ngates): does this API make it too easy to forget about sliced data?
+    pub fn into_owner<T>(self) -> Option<Arc<T>>
+    where
+        T: Send + Sync + 'static,
+    {
+        self.owner.as_any().downcast::<T>().ok()
+    }
+
+    /// Attempts to convert this buffer into a mutable buffer if there is only a single
+    /// reference to the buffer. Otherwise, returns the original buffer.
+    pub fn try_into_mut(self) -> Result<BufferMut, Self> {
+        self.owner
+            .into_mut()
+            .map(|owner_mut| BufferMut {
+                owner: owner_mut.owner,
+                ptr: self.ptr as *mut u8,
+                length: self.length,
+            })
+            .map_err(|owner| Self {
+                owner,
+                ptr: self.ptr,
+                length: self.length,
+            })
+    }
+
+    /// Returns a slice of self for the provided range.
+    /// This operation is `O(1)`.
+    pub fn slice(&self, offset: usize) -> Self {
+        let mut sliced = self.clone();
+
+        assert!(
+            offset <= sliced.length,
+            "the offset of the new Buffer cannot exceed the existing length: offset={} length={}",
+            offset,
+            sliced.length
+        );
+        sliced.length -= offset;
+        // SAFETY: `offset <= self.length`
+        sliced.ptr = unsafe { sliced.ptr.add(offset) };
+        sliced
+    }
+
+    /// Returns a new [Buffer] that is a slice of this buffer starting at `offset`, with
+    /// `length` bytes.
+    ///
+    /// # Panics
+    ///
+    /// Panics iff `(offset + length)` is larger than the existing length.
+    pub fn slice_with_length(&self, offset: usize, length: usize) -> Self {
```

Comment:
offset + length is always easily confused with start + end... wonder if we should just take a Range<usize>?

Reply:
Not with a function named slice_with_length! 

Range is annoying, because it forces you to specify an upper bound.

Reply:
`slice(&self, range: impl std::ops::RangeBounds<usize>)`